ext/drx/scatter_gather_aarch64.c

```diff
@@ -369,57 +423,39 @@ expand_scalar_plus_vector(void *drcontext, instrlist_t *bb, instr_t *sg_instr,
         EMIT(lastb_sve_scalar, opnd_create_reg(scratch_gpr0),
              opnd_create_reg(scratch_pred),
              opnd_create_reg_element_vector(sg_info->index_reg, sg_info->element_size));
+        const reg_id_t scalar_index = scratch_gpr0;
 
-        opnd_t mem = opnd_create_base_disp_shift_aarch64(
-            sg_info->base_reg, scratch_gpr0, sg_info->extend, sg_info->scaled,
-            /*disp=*/0, /*flags=*/0, sg_info->scalar_value_size, sg_info->extend_amount);
         if (sg_info->is_load) {
-            /* ldr[bh]  scratch_gpr0, [base_reg, scratch_gpr0, mod #amount] */
-            if (sg_info->is_scalar_value_signed) {
-                const reg_id_t ld_dst =
-                    reg_resize_to_opsz(scratch_gpr0, sg_info->element_size);
-                switch (sg_info->scalar_value_size) {
-                case OPSZ_1: EMIT(ldrsb, opnd_create_reg(ld_dst), mem); break;
-                case OPSZ_2: EMIT(ldrsh, opnd_create_reg(ld_dst), mem); break;
-                case OPSZ_4: EMIT(ldrsw, opnd_create_reg(ld_dst), mem); break;
-                default: DR_ASSERT_MSG(false, "Invalid scatter_gather_info_t data");
-                }
-            } else {
-                const reg_id_t scratch_gpr0_w = reg_resize_to_opsz(scratch_gpr0, OPSZ_4);
-                switch (sg_info->scalar_value_size) {
-                case OPSZ_1: EMIT(ldrb, opnd_create_reg(scratch_gpr0_w), mem); break;
-                case OPSZ_2: EMIT(ldrh, opnd_create_reg(scratch_gpr0_w), mem); break;
-                case OPSZ_4: EMIT(ldr, opnd_create_reg(scratch_gpr0_w), mem); break;
-                case OPSZ_8: EMIT(ldr, opnd_create_reg(scratch_gpr0), mem); break;
-                default: DR_ASSERT_MSG(false, "Invalid scatter_gather_info_t data");
-                }
-            }
+            const reg_id_t scalar_dst = scratch_gpr0;
+
+            /* ldr[bh]  scalar_dst, [mem] */
+            instrlist_preinsert(
```

Comment:
nit: optional: sg_info->is_load is reevaluated inside create_scalar_load_or_store. I wonder if it'll more any more readable/simpler if we split create_scalar_load_or_store into create_scalar_load and create_scalar_store.

Reply:
I looked at this but I think its simpler as it is now.